actors/ppo_atari_pz.py

```diff
@@ -291,84 +337,106 @@ async def trial_sample_sequences_producer_impl(self, sample_producer_session: Sa
             obs_flat = player_observation_space.deserialize(actor_sample.observation).flat_value
-            observation_value = torch.unsqueeze(
-                torch.permute(torch.tensor(obs_flat, dtype=self._dtype).reshape(obs_shape), (2, 0, 1)), dim=0
+            obs_tensor = torch.tensor(obs_flat, dtype=self._dtype).reshape(obs_shape).clone()
+            observation_value = torch.unsqueeze(torch.permute(obs_tensor, (2, 0, 1)), dim=0)
+            done = torch.ones(1, dtype=torch.float32) if trial_done else torch.zeros(1, dtype=torch.float32)
+            reward_value = (
+                torch.tensor(actor_sample.reward, dtype=self._dtype)
+                if actor_sample.reward is not None
+                else torch.tensor(0, dtype=self._dtype)
             )
-            observation.append(observation_value)
-            if sample.trial_state == cogment.TrialState.ENDED:
-                done.append(torch.ones(1, dtype=self._dtype))
-                break
 
-            action_value = torch.tensor(player_action_space.deserialize(actor_sample.action).value, dtype=self._dtype)
-            reward_value = torch.tensor(
-                actor_sample.reward if actor_sample.reward is not None else 0, dtype=self._dtype
-            )
-            action.append(action_value)
-            reward.append(reward_value)
-            done.append(torch.zeros(1, dtype=self._dtype))
-            current_players.append(player_actor_name)
+            if not trial_done:
+                action_value = torch.tensor(
+                    player_action_space.deserialize(actor_sample.action).value, dtype=self._dtype
+                )
+                current_players.append(player_actor_name)
 
-        # Keeping the samples grouped by trial by emitting only one grouped sample at the end of the trial
-        sample_producer_session.produce_sample((observation, action, reward, done, current_players))
+            # Compute values and log probs
+            if step % self._cfg.num_rollout_steps < self._cfg.num_rollout_steps and not trial_done:
+                with torch.no_grad():
+                    value, log_prob, _ = model.network.get_action_value(
+                        observation=observation_value, action=action_value
+                    )
+                    values.append(value.squeeze(0).cpu())
+                    log_probs.append(log_prob.squeeze(0).cpu())
+
+                # Add sample to rollout replay buffer
+                rollout_buffer.add(observation=observation_value, action=action_value, reward=reward_value, done=done)
+
+            # Save episode reward i.e., number of total steps for an episode
+            step += 1
+            total_reward += 1
+            if trial_done:
+                episode_rewards.append(torch.tensor(total_reward / num_players, dtype=self._dtype))
+                total_reward = 0
+
+            # Produce sample for training task
+            if step % self._cfg.num_rollout_steps == 0 or trial_done:
```

Comment:
We only want to produce after N steps of rollout, In this case will be 128 (i.e., `num_rollout_steps: 128`). This is one of a feature of PPO algorithm (see [Algorithm 1](https://arxiv.org/pdf/1707.06347.pdf)). The reason is that the number of steps of a pong trial is >3000 steps. We want to update the algorithm each small N steps in order to avoid the divergence in performance. For example, in the case of Pong, 128 is commonly used. For hopper, it is 2048